### PR TITLE
add the ability to specify a version number for a tracing library.

### PR DESCRIPTION
**Pull request by jonbodner**
- add the ability to specify a version number for a tracing library using a : after the langauge name.
- remove datadog-apm-library-all, because it causes problems when you want to force an update to a latest version of a tracing library.

https://datadoghq.atlassian.net/browse/APMON-355

### CODE REVIEW

**Review by Kyle-Verhoog — COMMENTED** [review]
probably worth splitting the changes for removing the all package and introducing the library versions since they're unrelated

**Comment by jonbodner** [comment]
> probably worth splitting the changes for removing the all package and introducing the library versions since they're unrelated

fair point. I'll split it up.

File: install_script.sh.template
Comment (bkabrda) [inline install_script.sh.template:445]: The `all` shouldn't be here

File: install_script.sh.template
Comment (bkabrda) [inline install_script.sh.template:426]: We don't want this clause any more as we're removing "all"

**Review by bkabrda — APPROVED** [review]
LGTM :rocket:

**Comment by bkabrda** [comment]
Actually one more comment - can we uninstall the `datadog-apm-library-all` explicitly as we want to move away from using it? It would be good to get it uninstalled from customer systems.

**Comment by jonbodner** [comment]
Telemetry has been added since the original approval. Can we get a new approval now? I've check with @yshapiro-57 to make sure that the telemetry is being received correctly.

**Comment by bkabrda** [comment]
I actually realized that there is one problem with this - if the installed versions are non-latest, the next `apt upgrade`/`dnf update` will upgrade them to latest versions. I guess this is not what we want here, right? If we want to avoid that, we would need to use locking mechanisms:

* https://manpages.ubuntu.com/manpages/focal/en/man8/dpkg-hold.8.html
* https://linux.die.net/man/5/yum.conf (search for `includepkgs`) - this would be a little more tricky, because it requires changing the actual repofile and listing all packages that are being installed, not just the tracer libraries.

Let me know if you need some help with this

**Comment by jonbodner** [comment]
> I actually realized that there is one problem with this - if the installed versions are non-latest, the next `apt upgrade`/`dnf update` will upgrade them to latest versions. I guess this is not what we want here, right? If we want to avoid that, we would need to use locking mechanisms:
> 
> * https://manpages.ubuntu.com/manpages/focal/en/man8/dpkg-hold.8.html
> * https://linux.die.net/man/5/yum.conf (search for `includepkgs`) - this would be a little more tricky, because it requires changing the actual repofile and listing all packages that are being installed, not just the tracer libraries.
> 
> Let me know if you need some help with this

Hrm, that's a good point. We would also want to unlock when the script runs so that the script can perform the update.

How much work is it to get this added to the install script?

**Comment by bkabrda** [comment]
It's not much work, I would say maybe 1 day including testing. For RPM, the repofile gets rewritten before the "yum" call is issued, so that will take care of updating the lock and things will work fine AFAICS. For DEB, you're right that we'll first need to unlock, then install and then lock again.

**Comment by bkabrda** [comment]
@jonbodner I pushed a commit that implements the pinning - it'd be great if you could test it.

One caveat: with yum (CentOS/RHEL 6 and 7), if you pin a tracer and then try to rerun the script with a pin of lower version (so when you try to downgrade), the tracer will not be downgraded. This is a known yum behavior and would have to be worked around by uninstalling the tracer and then installing it. I think we can probably live with this, but I will leave it up to you to decide whether we want to invest more time in fixing this.

**Review by spencergilbert — APPROVED** [review]

File: install_script.sh.template
Comment (chouquette) [inline install_script.sh.template:132]: it feels like this block could be simplified by matching on the last segment on the string, something like this:
```
function get_apm_lib {
   case $1 in
  datadog-apm-library-*)
    case $1 in
      *-java|*-python|*-dotnet|*-ruby)
          echo ${1#datadog-apm-library}
          ;;
      *-js)
          echo nodejs
          ;;
      *)
         fail
    esac
     ;;
  *)
   fail
   ;;
esac
}

...

library_telemetry_info="${library_telemetry_info},\"specified_$(get_apm_lib $cur_lib)_lib_version\": \"${cur_version}\""
```

The same function could be reused in the similar block below

File: install_script.sh.template
Comment (jonbodner) [reply install_script.sh.template]: resolved

File: install_script.sh.template
Comment (jonbodner) [reply install_script.sh.template]: resolved

**Pull request merged**
